src/Workspaces/Core/Portable/Workspace/WorkspaceRegistration.cs

```diff
@@ -23,12 +24,23 @@ internal void SetWorkspaceAndRaiseEvents(Workspace workspace)
 
         internal void SetWorkspace(Workspace workspace)
         {
-            _registeredWorkspace = workspace;
+            Workspace = workspace;
         }
 
         internal void RaiseEvents()
         {
-            WorkspaceChanged?.Invoke(this, EventArgs.Empty);
+            lock (_gate)
```

Comment:
Could this possibly cause any deadlock, especially when the lock might be acquired on UI thread?

Reply:
I can't think of an obvious way it would happen. It seems you'd more or less have to have code under this lock modifying another Workspace or text buffer, and that one then simultaneously trying to update the registration of the original one again.

In any case, if I'm wrong, the dump will be easy enough to see what happened!

Reply:
@genlu yep. it has theoretical deadlock potential since the event is public and we can't control what the other side will do.

the linked github issue lists possible other approaches, if you have an idea, please list there other approaches we could take.

Reply:
well, we could queue the events and then fire each one only once the previous had completed.  is there a good reason not to do that?

Reply:
At this point, largely risk if we're taking this as a hotfix: it's unclear what expectations there might exist on UnregisterText raising this immediately.

If we did a queue, it's not entirely clear to me which thread that queue is associated with, unfortunately.